### PR TITLE
Feature/fork deploy and migration

### PR DESCRIPTION
**Pull request by horodnicdragos**

### CODE REVIEW

File: contracts/mocks/Disposables/DisposableCover.sol
Comment (danoctavian) [inline contracts/mocks/Disposables/DisposableCover.sol:45]: spelling error here: initialProuctsCount -> initialProductsCount

Comment (danoctavian) [reply contracts/mocks/Disposables/DisposableCover.sol:45]: why is it upserted ?

You upsert if you may also update an existing record, but here it just pushes one at the end, so it would be just ProductAdded

File: contracts/mocks/Disposables/DisposableCover.sol
Comment (danoctavian) [inline contracts/mocks/Disposables/DisposableCover.sol:56]: same question as above with regards to `Upserted`

File: contracts/interfaces/IStakingPool.sol
Comment (horodnicdragos) [reply contracts/interfaces/IStakingPool.sol:24]: It should be 0-100, each unit representing one percent. @shark0der let us know if anything is about to change here.

File: contracts/modules/capital/Pool.sol
Comment (danoctavian) [inline contracts/modules/capital/Pool.sol:108]: these used to be parameters.

This is definitely nice for clarity of what's being used, but it might make some tests cumbersome if we want to the test the Pool with different values.

File: contracts/modules/capital/Pool.sol
Comment (danoctavian) [inline contracts/modules/capital/Pool.sol:493]: why change the comment style on top of functions?

the /** */ is pretty standard across multiple languages and looks better/more familiar for most devs.

File: contracts/mocks/Disposables/DisposableCover.sol
Comment (horodnicdragos) [reply contracts/mocks/Disposables/DisposableCover.sol:45]: The event is emitted whenever a product changes. This means both adding a new product and editing an existing one. It's purpose is to follow changes to any specific wordings similar to what we currently have in the UI as "excluded scenarios" which vary from one product to another. This information is stored on IPFS and the event is a reference to the cid referencing the file that contains it.

File: contracts/modules/capital/Pool.sol
Comment (horodnicdragos) [reply contracts/modules/capital/Pool.sol:108]: We have setSwapDetails which can easily be used to test different values.

File: contracts/modules/capital/Pool.sol
Comment (horodnicdragos) [reply contracts/modules/capital/Pool.sol:493]: Here's why:

1. You can easily copy or move chunks of the comment without having to worry about turning it back into a comment.
2. It doesn't make any assumptions about smart tooling that handles comments.
3. It's 2 lines smaller.
4. It's also standard across multiple languages.
5. It doesn't affect the code or other devs in any way, even if we're looking in subjective matters such as looks and familiarity. You don't have to stick to it, but it definitely makes my dev experience more pleasent.

File: contracts/modules/staking/PooledStaking.sol
Comment (danoctavian) [inline contracts/modules/staking/PooledStaking.sol:701]: is this still needed in this commented out form?

File: contracts/modules/staking/PooledStaking.sol
Comment (danoctavian) [inline contracts/modules/staking/PooledStaking.sol:960]: misspelling here i'm assuming 

getV1PercForProduct
getV1PercentageForProduct
getV1PriceForProduct

File: contracts/modules/staking/PooledStaking.sol
Comment (danoctavian) [inline contracts/modules/staking/PooledStaking.sol:1090]: are we going to keep updating these to match the latest number?

File: contracts/modules/staking/PooledStaking.sol
Comment (danoctavian) [inline contracts/modules/staking/PooledStaking.sol:1382]: we should measure gas here to see if there's any issue with max block size and whether we need iterations.

File: contracts/modules/staking/PooledStaking.sol
Comment (danoctavian) [inline contracts/modules/staking/PooledStaking.sol:1454]: so for addresses that are not in the initial migration list, the groupId is self-defined (parameter).

Is stakingPool.deposit supposed to revert if that groupId is wrong (not in the right range)?

File: contracts/modules/capital/Pool.sol
Comment (danoctavian) [reply contracts/modules/capital/Pool.sol:493]: Understood.

I would want things to be consistent so all comments in 1 way, i prefer the old style. I'll let the others chime in on this one.

File: contracts/modules/staking/PooledStaking.sol
Comment (horodnicdragos) [reply contracts/modules/staking/PooledStaking.sol:881]: Try taking a look at the diff as a whole. There were incrementations and decrementations that were cancelling each other out, sometimes ending up with underflows. 

That's one of the reasons why I left some things commented (unchecked operations which would be required to allow underflows). Thus you won't be able to minimize changes but instead pick one of the two approaches. The one involving underflows seemed fragile to me. 

The test coverage should be pretty good for this part of the code.

File: contracts/modules/staking/PooledStaking.sol
Comment (horodnicdragos) [reply contracts/modules/staking/PooledStaking.sol:1090]: Yes, `get-v1-cover-prices.js` does that automatically.

File: contracts/modules/staking/PooledStaking.sol
Comment (danoctavian) [inline contracts/modules/staking/PooledStaking.sol:1418]: now that i see this, should we leave the console.log in the deploy to have this logging?

Or just have a custom event, which adds less to the bytecode

File: contracts/modules/staking/PooledStaking.sol
Comment (danoctavian) [inline contracts/modules/staking/PooledStaking.sol:1464]: *deposit

File: contracts/modules/v2/CoverMigrator.sol
Comment (danoctavian) [inline contracts/modules/v2/CoverMigrator.sol:18]: is this replacing an existing contract?

File: contracts/modules/staking/PooledStaking.sol
Comment (horodnicdragos) [reply contracts/modules/staking/PooledStaking.sol:1454]: I think it should. At the moment it looks like it only checks whether the group is in the future.

`    require(groupId >= firstActiveGroupId, "StakingPool: Requested group has expired");`

Added a todo in StakingPool.sol.

File: contracts/modules/v2/CoverMigrator.sol
Comment (horodnicdragos) [reply contracts/modules/v2/CoverMigrator.sol:18]: Yes, Claims.sol. It's not explicitly mentioned in the comment above but it should be obvious. Some contracts use Claims.sol directly instead of Gateway.sol so we're leaving this CoverMigrator contract in place (CL) to allow cover migration for 1y. It can then be removed from Master entirely.

**Review by roxdanila — COMMENTED** [review]
Adding a few more comments to the mix. Still need to dive a bit more in the staking migration

File: scripts/create-testnet-version-data.js
Comment (horodnicdragos) [reply scripts/create-testnet-version-data.js:58]: I think we need that in the UI to display v1 claims.

File: contracts/modules/governance/MemberRoles.sol
Comment (horodnicdragos) [reply contracts/modules/governance/MemberRoles.sol:195]: Not sure if this was (or still is) a legal requirement. I'll remove it if it's no longer needed.

**Review by roxdanila — APPROVED** [review]

**Comment by roxdanila** [comment]
Please fix/skip failing tests before merging: https://app.circleci.com/pipelines/github/NexusMutual/smart-contracts/1381/workflows/b4bcda3d-5352-4821-bad7-9f976a949e8f/jobs/2817

File: contracts/modules/staking/PooledStaking.sol
Comment (shark0der) [reply contracts/modules/staking/PooledStaking.sol:883]: The underflow was intended. The patch looks correct to me and indeed, the tests should cover this. The edge case here and in the burn function as well is when there's a single staker left and his stake is zero so we're removing the entry from the `_contractStakers` array.

Comment (shark0der) [reply contracts/modules/staking/PooledStaking.sol:883]: processBurns.js: the "should remove stakers when burning 0-deposit stakers" test for example covers this edge-case in the burn function.

**Pull request merged**
